### DISCUSSION TITLE
How to correctly load images asynch?

### DISCUSSION THREAD

**Discussion by n00kii** [emilk/egui#1748] [Q&A] [answered]
First I tried doing it with `thread::spawn` and for some reason, the main thread would block while the worker threads were doing their thing. So now I'm trying to follow the example [here](https://github.com/emilk/egui/blob/master/examples/download_image/src/main.rs) and yet the main thread still blocks. I'm new to rust so forgive me for any dumbness but here's  the relevant code. 

My goal is to have a UI where, upon clicking scan, there will be two columns; one with labels of every file in the scanned directory, and  another column with the image previews. 
```rust
// a struct that is created for every directory entry when loading, holds information useful for the ui
struct MediaEntry {
    started_load: bool,
    attempted_load: bool,
    dir_entry: DirEntry,
    is_selected: bool,
    image: Option<Promise<Result<RetainedImage>>>
}

...

// below is called in update fn (the rest omitted for clarity)
// scan button was clicked, and scanned_dirs has loaded with mediaentries
if let Some(scanned_dirs) = &mut self.scanned_dir_entries {
    // iterate through each mediaentry to draw its name on the sidebar, and to load its image
    // wrapped in an arc mutex for multithreading purposes
    for media_entry_arc in scanned_dirs.iter_mut() {
        // take
        let media_entry_mutex = media_entry_arc.lock();
        if media_entry_mutex.is_err() {
            println!("skipped errored mutex");
            continue;
        }

        let mut media_entry = media_entry_mutex.unwrap();

        // display label stuff 
        let empty_path = Path::new("");
        let dir_entry = &media_entry.dir_entry;
        let dir_entry_path = dir_entry.path();
        let dir_entry_parent = dir_entry_path
            .parent()
            .unwrap_or(&empty_path)
            .file_name()
            .unwrap_or(empty_path.as_os_str())
            .to_str()
            .unwrap_or("");
        let dir_entry_filename = dir_entry_path
            .file_name()
            .unwrap_or(empty_path.as_os_str())
            .to_str()
            .unwrap_or("");
        let dir_entry_label = format!("{dir_entry_parent}/{dir_entry_filename}");

        // create label in left column
        if files_col
            .selectable_label(media_entry.is_selected, dir_entry_label)
            .clicked()
        {
            media_entry.is_selected = !media_entry.is_selected;
        };

        // promise for loading image
        let promise = media_entry.image.get_or_insert_with(|| {
            // clone the arc, make spawn a thread 
            let media_entry_arc_clone = Arc::clone(&media_entry_arc);
            let promise = Promise::spawn_thread(
                format!("load_{dir_entry_filename}"),
                move || {
                    // take mutex, then load image (slow)
                    let media_entry_mutex = media_entry_arc_clone.lock();
                    let mut media_entry = media_entry_mutex.unwrap();
                    media_entry.load_image()
                },
            );
            promise
        });
    }
}

```
As you can see, I'm not even displaying the image yet, and still the main thread halts. If it matters, looking in task manager reveals the process spikes in CPU usage, close to 100%, but I still can't tell what's going on

Thanks for any help

**Comment by dacid44** [comment, marked as answer]
I can see two main things that could be the issue:
- It doesn't appear to happen in this snippet, but wherever you're reading the mutex, make sure you're not blocking on waiting for it to lock. Ideally you check if it's locked, and if so, continue without the image and try again later.
- Another possibility is that you're deadlocked with the main thread holding the lock for the mutex, but for some reason won't drop it until after the worker thread joins, which won't happen until the worker thread can lock the mutex, which won't happen until... (etc.) I don't think this is as likely, but TLDR make sure mutexes are unlocked when they no longer need to be locked. This is often done implicitly but sometimes it's good to make sure.

Does the app ever actually unfreeze? If not, then you might have a different issue. As for the 100% usage, make sure you're only dispatching one worker thread per image. If this is run in `update()`, then there's a chance (depending on your app's logic) that you're dispatching a new thread per image, _per frame_.

I'd try throwing some print statements on there, maybe with some kind of IDs so you can tell which thread is printing what. That's often a good first step in debugging threading code.

Edit: I'll add an example from [one of my own projects](https://github.com/dacid44/android_apps_egui/blob/master/src/app.rs). The image fetching code is at the bottom. The implementation is different (tokio async tasks) but the idea is similar. Maybe it'll be useful. (Fair warning: old, messy code!)

**Reply by n00kii** [reply]
Your first point turned out to be the reason, thanks!
